### DISCUSSION TITLE
unable to install policy ::/0 === ::/0

### DISCUSSION THREAD

**Discussion by bvlarsen** [strongswan/strongswan#1010] [General]
Hi,

I have a Strongswan VPN_GW with Cisco IOS-XE 16.12.04 router spokes. When I try setting up route-based IPSec between spokes and vpn_gw I get the message "unable to install policy ::/0 === ::/0" when the second spoke tries to join the vpn_gw.

On the Cisco routers I use a virtual tunnel interface to set up IPv6 over IPv4 IPSec encryption. This limits me to use ::/0-::/0 policy on all routers and control traffic by routing.

I control Strongswan with swanctl.

OS is Google Cloud Ubuntu 20.04 image.
Strongswan version is 5.8.2, but I have compiled 5.9.5 to se if behavior is different. It is not.

Log outputs the following:
06[NET] received packet: from <peer_ip_A>[500] to <VPN_GW>[500] (386 bytes)
06[ENC] parsed IKE_SA_INIT request 0 [ SA KE No V V V V N(NATD_S_IP) N(NATD_D_IP) ]
06[IKE] received Cisco Delete Reason vendor ID
06[ENC] received unknown vendor ID: 43:49:53:43:4f:56:50:4e:2d:52:45:56:2d:30:32
06[ENC] received unknown vendor ID: 43:49:53:43:4f:2d:44:59:4e:41:4d:49:43:2d:52:4f:55:54:45
06[IKE] received Cisco FlexVPN Supported vendor ID
06[IKE] <peer_ip_A> is initiating an IKE_SA
06[CFG] selected proposal: IKE:AES_GCM_16_256/PRF_HMAC_SHA2_384/ECP_521
06[IKE] local host is behind NAT, sending keep alives
06[IKE] remote host is behind NAT
06[ENC] generating IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(CHDLESS_SUP) N(MULT_AUTH) ]
06[NET] sending packet: from <VPN_GW>[500] to <peer_ip_A>[500] (316 bytes)
09[NET] received packet: from <peer_ip_A>[4500] to <VPN_GW>[4500] (634 bytes)
09[ENC] unknown attribute type (28692)
09[ENC] parsed IKE_AUTH request 1 [ V IDi AUTH CPRQ(DNS DNS NBNS NBNS SUBNET DNS6 SUBNET6 VER U_SPLITDNS U_BANNER (28692) U_BKPSRV U_DEFDOM) SA TSi TSr N(INIT_CONTACT) N(SET_WINSIZE) N(ESP_TFC_PAD_N) N(NON_FIRST_FRAG) ]
09[CFG] looking for peer configs matching <VPN_GW>[%any]...<peer_ip_A>[net2.com]
09[CFG] selected peer config 'net2sa'
09[IKE] authentication of 'net2.com' with pre-shared key successful
09[IKE] received ESP_TFC_PADDING_NOT_SUPPORTED, not using ESPv3 TFC padding
09[IKE] authentication of 'vpngw.com' (myself) with pre-shared key
09[IKE] IKE_SA net2sa[25] established between <VPN_GW>[vpngw.com]...<peer_ip_A>[net2.com]
09[IKE] scheduling reauthentication in 10730s
09[IKE] maximum IKE_SA lifetime 11810s
09[CFG] selected proposal: ESP:AES_GCM_16_256/NO_EXT_SEQ
09[KNL] got SPI c5aedb8b
09[CHD] CHILD_SA net2{21} state change: CREATED => INSTALLING
09[CHD]   using AES_GCM_16 for encryption
09[CHD] adding inbound ESP SA
09[CHD]   SPI 0xc5aedb8b, src <peer_ip_A> dst <VPN_GW>
09[KNL] adding SAD entry with SPI c5aedb8b and reqid {2}
09[KNL]   using encryption algorithm AES_GCM_16 with key size 288
09[KNL]   using replay window of 32 packets
09[KNL]   HW offload: no
09[CHD] adding outbound ESP SA
09[CHD]   SPI 0xb875b45a, src <VPN_GW> dst <peer_ip_A>
09[KNL] adding SAD entry with SPI b875b45a and reqid {2} (mark 120/0xffffffff)
09[KNL]   using encryption algorithm AES_GCM_16 with key size 288
09[KNL]   using replay window of 0 packets
09[KNL]   HW offload: no
09[CFG] unable to install policy ::/0 === ::/0 in for reqid 2, the same policy for reqid 1 exists
09[CFG] unable to install policy ::/0 === ::/0 fwd for reqid 2, the same policy for reqid 1 exists
09[KNL] adding policy ::/0 === ::/0 out (mark 120/0xffffffff) [priority 399999, refcount 1]
09[IKE] unable to install IPsec policies (SPD) in kernel
09[IKE] failed to establish CHILD_SA, keeping IKE_SA
09[CHD] CHILD_SA net2{21} state change: INSTALLING => DESTROYING
09[KNL] deleting policy ::/0 === ::/0 out (mark 120/0xffffffff)
09[KNL] deleting policy ::/0 === ::/0 in
09[KNL] policy still used by another CHILD_SA, not removed
09[KNL] not updating policy ::/0 === ::/0 in [priority 399999, refcount 1]
09[KNL] deleting policy ::/0 === ::/0 fwd
09[KNL] policy still used by another CHILD_SA, not removed
09[KNL] not updating policy ::/0 === ::/0 fwd [priority 399999, refcount 1]
09[KNL] deleting SAD entry with SPI c5aedb8b
09[KNL] deleted SAD entry with SPI c5aedb8b
09[KNL] deleting SAD entry with SPI b875b45a (mark 120/0xffffffff)
09[KNL] deleted SAD entry with SPI b875b45a (mark 120/0xffffffff)
09[ENC] generating IKE_AUTH response 1 [ IDr AUTH N(TS_UNACCEPT) ]
09[NET] sending packet: from <VPN_GW>[4500] to <peer_ip_A>[4500] (138 bytes)
10[NET] received packet: from <peer_ip_A>[4500] to <VPN_GW>[4500] (69 bytes)
10[ENC] parsed INFORMATIONAL request 2 [ D ]

swanctl --list-conn, only lists the first router that joined:
net3sa: #5, ESTABLISHED, IKEv2, 32eb9e70e1c74bdc_i d52ba646fce9b7b2_r*
  local  'vpngw.com' @ <VPN_GW>[4500]
  remote 'net3.com' @ <peer_ip_B>[27011]
  AES_GCM_16-256/PRF_HMAC_SHA2_384/ECP_521
  established 3855s ago, reauth in 6893s
  net3: #84, reqid 1, INSTALLED, TUNNEL-in-UDP, ESP:AES_GCM_16-256/ECP_521
    installed 606s ago, rekeying in 2976s, expires in 3355s
    in  c1127b17 (-|0x0000000a),   9196 bytes,   121 packets,     2s ago
    out f8297495 (0x00000079|0x0000000a),      0 bytes,     0 packets
    local  ::/0
    remote ::/0

swancrl -L:
net2sa: IKEv2, reauthentication every 10800s, no rekeying
  local:  %any
  remote: %any
  local pre-shared key authentication:
    id: vpngw.com
  remote pre-shared key authentication:
    id: net2.com
  net2: TUNNEL, rekeying every 3600s
    local:  ::/0
    remote: ::/0
net3sa: IKEv2, reauthentication every 10800s, no rekeying
  local:  %any
  remote: %any
  local pre-shared key authentication:
    id: vpngw.com
  remote pre-shared key authentication:
    id: net3.com
  net3: TUNNEL, rekeying every 3600s
    local:  ::/0
    remote: ::/0

peer net2sa (peer_ip_A, in log):
connections {
  net2sa {
    local_addrs  = %any
    remote_addrs = %any
    local {
      auth = psk
      id = vpngw.com
    }
    remote {
      auth = psk
      id = net2.com
    }
    children {
      net2 {
        mode = tunnel
        local_ts  = ::/0
        remote_ts = ::/0
        if_id_in = 10
        if_id_out = 10
        esp_proposals = aes256gcm128-prfsha384-ecp521
        updown = <script>
        set_mark_in = 120
        mark_out = 120
      }
    }
    version = 2
    mobike = no
    reauth_time = 10800
    proposals = aes256gcm128-prfsha384-ecp521
  }
}
secrets {
  ike-net2 {
    id-1 = net2.com
    id-2 = vpngw.com
    secret = <secret>
  }
}

peer net3sa:
connections {
  net3sa {
    local_addrs  = %any
    remote_addrs = %any
    local {
      auth = psk
      id = vpngw.com
    }
    remote {
      auth = psk
      id = net3.com
    }
    children {
      net3 {
        mode = tunnel
        local_ts  = ::/0
        remote_ts = ::/0
        if_id_in = 10
        if_id_out = 10
        esp_proposals = aes256gcm128-prfsha384-ecp521
        updown = <script>
        set_mark_in = 121
        mark_out = 121
      }
    }
    version = 2
    mobike = no
    reauth_time = 10800
    proposals = aes256gcm128-prfsha384-ecp521
  }
}
secrets {
  ike-net3 {
    id-1 = net3.com
    id-2 = vpngw.com
    secret = <secret>
  }
}

As you can see the XFRM interface is the same for all connections. Different connections use different marks. I have tried slightly different, but largely overlapping selectors, and this works. So routing don't seem to be a problem. It seems the issue is that multiple connections use the exact same selectors.

This seems also to apply for 0.0.0.0/0-0.0.0.0/0.

<peer_ip_A> is not equal to <peer_ip_B>.

**Comment by tobiasbrunner** [comment]
> Different connections use different marks.

No, they don't. You configured 'set_mark_in` but I guess you meant to configure `mark_in` (or maybe both?). Check the documentation for the difference between these two options.

**Reply by bvlarsen** [reply]
Thank you for the response.

I have read the doc´s and tried different combinations, but got a bit lost I guess.
For a while I also tried:

        mark_in_sa = yes
        mark_in = 120

But that failed. A diagram similar to the linux kernel packet flow diagram would be very helpful.

Strongswan is great software!

**Reply by tobiasbrunner** [reply]
What do you mean? That's exactly the diagram that applies.
